src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -293,6 +433,132 @@ internal ProjectDependencyGraph WithAdditionalProjectReferences(ProjectId projec
+                    // reverse transitive references. Without more complicated checks, we cannot rule out the
+                    // possibility that A may have been removed from the reverse transitive references of D by the
+                    // removal of the A->B reference.
+                    continue;
+                }
+
+                if (references.Contains(projectId))
+                {
+                    builder.Remove(project);
+                }
+            }
+
+            // Invalidate the transitive references from the previously-referenced project
+            builder.Remove(referencedProjectId);
+
+            return builder.ToImmutable();
+        }
+
+        internal ProjectDependencyGraph WithProjectRemoved(ProjectId projectId)
+        {
+            Contract.ThrowIfFalse(_projectIds.Contains(projectId));
+
+            // Project ID set and direct forward references are trivially updated by removing the key corresponding to
+            // the project getting removed.
+            var projectIds = _projectIds.Remove(projectId);
+            var referencesMap = ComputeNewReferencesMapForRemovedProject(_lazyReverseReferencesMap, _referencesMap, projectId);
+
+            // The direct reverse references map is updated by removing the key for the project getting removed, and
+            // also updating any direct references to the removed project.
+            var reverseReferencesMap = _lazyReverseReferencesMap is object
+                ? ComputeNewReverseReferencesMapForRemovedProject(_referencesMap, _lazyReverseReferencesMap, projectId)
+                : null;
+            var transitiveReferencesMap = ComputeNewTransitiveReferencesMapForRemovedProject(_transitiveReferencesMap, projectId);
+            var reverseTransitiveReferencesMap = ComputeNewReverseTransitiveReferencesMapForRemovedProject(_reverseTransitiveReferencesMap, projectId);
+            return new ProjectDependencyGraph(
+                projectIds,
+                referencesMap,
+                reverseReferencesMap,
+                transitiveReferencesMap,
+                reverseTransitiveReferencesMap,
+                topologicallySortedProjects: default,
+                dependencySets: default);
+        }
+
+        internal ProjectDependencyGraph WithProjectReferenceRemoved(ProjectId projectId, ProjectId referencedProjectId)
+        {
+            Contract.ThrowIfFalse(_projectIds.Contains(projectId));
+            Contract.ThrowIfFalse(_referencesMap[projectId].Contains(referencedProjectId));
+
+            var projectIds = _projectIds;
+            var referencesMap = _referencesMap.SetItem(projectId, _referencesMap[projectId].Remove(referencedProjectId));
+            var reverseReferencesMap = _lazyReverseReferencesMap is object
+                ? ComputeNewReverseReferencesMapForRemovedProjectReference(_lazyReverseReferencesMap, projectId, referencedProjectId)
+                : null;
```

Comment:
could this check move *into* `ComputeNewReverseReferencesMapForRemovedProjectReference`?  the steps would look a lot cleaner. i.e. three steps where we call into a helper to compute the new map we need.

Reply:
applies to above method as well.

Reply:
See https://github.com/dotnet/roslyn/pull/40358/files#r369840076